pkg/safe/state.go

```diff
@@ -0,0 +1,110 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package safe
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/state"
+)
+
+// StateGet is a type safe wrapper around state.Get.
+func StateGet[T resource.Resource](ctx context.Context, st state.State, ptr resource.Pointer, options ...state.GetOption) (T, error) { //nolint:ireturn
+	got, err := st.Get(ctx, ptr, options...)
+	if err != nil {
+		var zero T
+
+		return zero, err
+	}
+
+	result, ok := got.(T)
+	if !ok {
+		var zero T
+
+		return zero, fmt.Errorf("type mismatch: expected %T, got %T", result, got)
+	}
+
+	return result, nil
+}
+
+// StateUpdateWithConflicts is a type safe wrapper around state.UpdateWithConflicts.
+func StateUpdateWithConflicts[T resource.Resource](ctx context.Context, st state.State, ptr resource.Pointer, updateFn func(T) error, options ...state.UpdateOption) (T, error) { //nolint:ireturn
+	got, err := st.UpdateWithConflicts(ctx, ptr, func(r resource.Resource) error {
+		arg, ok := r.(T)
+		if !ok {
+			return fmt.Errorf("type mismatch: expected %T, got %T", arg, r)
+		}
+
+		return updateFn(arg)
+	}, options...)
+	if err != nil {
+		var zero T
+
+		return zero, err
+	}
+
+	result, ok := got.(T)
+	if !ok {
+		var zero T
+
+		return zero, fmt.Errorf("type mismatch: expected %T, got %T", result, got)
+	}
+
+	return result, nil
+}
+
+// StateList is a type safe wrapper around state.List.
+func StateList[T resource.Resource](ctx context.Context, st state.State, ptr resource.Pointer, options ...state.ListOption) (Iterator[T], error) {
+	got, err := st.List(ctx, ptr, options...)
+	if err != nil {
+		var zero Iterator[T]
+
+		return zero, err
+	}
+
+	if len(got.Items) == 0 {
+		var zero Iterator[T]
+
+		return zero, nil
+	}
+
+	// Early assertion to make sure we don't have a type mismatch.
+	if _, ok := got.Items[0].(T); !ok {
+		var zero Iterator[T]
+
+		return zero, fmt.Errorf("type mismatch on the first element: expected %T, got %T", got.Items[0], got)
+	}
+
+	return NewIterator[T](got.Items), nil
+}
+
+// Iterator is a generic iterator over resource.Resource slice.
+type Iterator[T any] struct {
+	res []resource.Resource
+	pos int
+}
+
+// NewIterator returns a new iterator over the given slice.
+func NewIterator[T any](res []resource.Resource) Iterator[T] {
```

Comment:
return from `List()` a typed resource slice, and make `Iterator`

Reply:
Done.